# 1. Required Libraries:

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import itertools

import warnings
# Suppressing the  warnings
warnings.filterwarnings("ignore")

# 2. Importing the Data:

In [2]:
Sales_dataset = pd.read_csv("Pizza_Sale - pizza_sales.csv")
Ingredients_dataset = pd.read_csv("Pizza_ingredients - Pizza_ingredients.csv")

In [3]:
Sales_dataset 

pizza_id  order_id  pizza_name_id  quantity  order_date order_time  \
0             1         1     hawaiian_m         1    1/1/2015   11:38:36   
1             2         2  classic_dlx_m         1    1/1/2015   11:57:40   
2             3         2  five_cheese_l         1    1/1/2015   11:57:40   
3             4         2    ital_supr_l         1    1/1/2015   11:57:40   
4             5         2     mexicana_m         1    1/1/2015   11:57:40   
...         ...       ...            ...       ...         ...        ...   
48615     48616     21348  ckn_alfredo_m         1  31-12-2015   21:23:10   
48616     48617     21348  four_cheese_l         1  31-12-2015   21:23:10   
48617     48618     21348   napolitana_s         1  31-12-2015   21:23:10   
48618     48619     21349     mexicana_l         1  31-12-2015   22:09:54   
48619     48620     21350      bbq_ckn_s         1  31-12-2015   23:02:05   

       unit_price  total_price pizza_size pizza_category  \
0           13.25        13.25          M        Classic   
1           16.00        16.00          M        Classic   
2           18.50        18.50          L         Veggie   
3           20.75        20.75          L        Supreme   
4           16.00        16.00          M         Veggie   
...           ...          ...        ...            ...   
48615       16.75        16.75          M        Chicken   
48616       17.95        17.95          L         Veggie   
48617       12.00        12.00          S        Classic   
48618       20.25        20.25          L         Veggie   
48619       12.75        12.75          S        Chicken   

                                       pizza_ingredients  \
0               Sliced Ham, Pineapple, Mozzarella Cheese   
1      Pepperoni, Mushrooms, Red Onions, Red Peppers,...   
2      Mozzarella Cheese, Provolone Cheese, Smoked Go...   
3      Calabrese Salami, Capocollo, Tomatoes, Red Oni...   
4      Tomatoes, Red Peppers, Jalapeno Peppers, Red O...   
...                                                  ...   
48615  Chicken, Red Onions, Red Peppers, Mushrooms, A...   
48616  Ricotta Cheese, Gorgonzola Piccante Cheese, Mo...   
48617  Tomatoes, Anchovies, Green Olives, Red Onions,...   
48618  Tomatoes, Red Peppers, Jalapeno Peppers, Red O...   
48619  Barbecued Chicken, Red Peppers, Green Peppers,...   

                       pizza_name  
0              The Hawaiian Pizza  
1        The Classic Deluxe Pizza  
2           The Five Cheese Pizza  
3       The Italian Supreme Pizza  
4              The Mexicana Pizza  
...                           ...  
48615   The Chicken Alfredo Pizza  
48616       The Four Cheese Pizza  
48617        The Napolitana Pizza  
48618          The Mexicana Pizza  
48619  The Barbecue Chicken Pizza  

[48620 rows x 12 columns]

In [4]:
Ingredients_dataset

pizza_name_id                         pizza_name  pizza_ingredients  \
0       bbq_ckn_l         The Barbecue Chicken Pizza  Barbecued Chicken   
1       bbq_ckn_l         The Barbecue Chicken Pizza        Red Peppers   
2       bbq_ckn_l         The Barbecue Chicken Pizza      Green Peppers   
3       bbq_ckn_l         The Barbecue Chicken Pizza           Tomatoes   
4       bbq_ckn_l         The Barbecue Chicken Pizza         Red Onions   
..            ...                                ...                ...   
513  veggie_veg_s  The Vegetables + Vegetables Pizza      Green Peppers   
514  veggie_veg_s  The Vegetables + Vegetables Pizza         Red Onions   
515  veggie_veg_s  The Vegetables + Vegetables Pizza           Zucchini   
516  veggie_veg_s  The Vegetables + Vegetables Pizza            Spinach   
517  veggie_veg_s  The Vegetables + Vegetables Pizza             Garlic   

     Items_Qty_In_Grams  
0                  40.0  
1                  15.0  
2                  20.0  
3                  30.0  
4                  60.0  
..                  ...  
513                20.0  
514                20.0  
515                10.0  
516                15.0  
517                 5.0  

[518 rows x 4 columns]

# Dataset Overview:

1) Pizza_Sale.csv with 48,620 rows and 12 columns.
2) Pizza_Ingredients.csv with 518 rows and 4 columns.

# Sales Data

In [5]:
Sales_dataset.head()

pizza_id  order_id  pizza_name_id  quantity order_date order_time  \
0         1         1     hawaiian_m         1   1/1/2015   11:38:36   
1         2         2  classic_dlx_m         1   1/1/2015   11:57:40   
2         3         2  five_cheese_l         1   1/1/2015   11:57:40   
3         4         2    ital_supr_l         1   1/1/2015   11:57:40   
4         5         2     mexicana_m         1   1/1/2015   11:57:40   

   unit_price  total_price pizza_size pizza_category  \
0       13.25        13.25          M        Classic   
1       16.00        16.00          M        Classic   
2       18.50        18.50          L         Veggie   
3       20.75        20.75          L        Supreme   
4       16.00        16.00          M         Veggie   

                                   pizza_ingredients  \
0           Sliced Ham, Pineapple, Mozzarella Cheese   
1  Pepperoni, Mushrooms, Red Onions, Red Peppers,...   
2  Mozzarella Cheese, Provolone Cheese, Smoked Go...   
3  Calabrese Salami, Capocollo, Tomatoes, Red Oni...   
4  Tomatoes, Red Peppers, Jalapeno Peppers, Red O...   

                  pizza_name  
0         The Hawaiian Pizza  
1   The Classic Deluxe Pizza  
2      The Five Cheese Pizza  
3  The Italian Supreme Pizza  
4         The Mexicana Pizza

In [6]:
Sales_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48620 entries, 0 to 48619
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   pizza_id           48620 non-null  int64  
 1   order_id           48620 non-null  int64  
 2   pizza_name_id      48604 non-null  object 
 3   quantity           48620 non-null  int64  
 4   order_date         48620 non-null  object 
 5   order_time         48620 non-null  object 
 6   unit_price         48620 non-null  float64
 7   total_price        48613 non-null  float64
 8   pizza_size         48620 non-null  object 
 9   pizza_category     48597 non-null  object 
 10  pizza_ingredients  48607 non-null  object 
 11  pizza_name         48613 non-null  object 
dtypes: float64(2), int64(3), object(7)
memory usage: 4.5+ MB


In [7]:
Sales_dataset.describe()

pizza_id      order_id      quantity    unit_price   total_price
count  48620.000000  48620.000000  48620.000000  48620.000000  48613.000000
mean   24310.500000  10701.479761      1.019622     16.494132     16.821494
std    14035.529381   6180.119770      0.143077      3.621789      4.437520
min        1.000000      1.000000      1.000000      9.750000      9.750000
25%    12155.750000   5337.000000      1.000000     12.750000     12.750000
50%    24310.500000  10682.500000      1.000000     16.500000     16.500000
75%    36465.250000  16100.000000      1.000000     20.250000     20.500000
max    48620.000000  21350.000000      4.000000     35.950000     83.000000

# 4. Handling Missing Values in Sales Data:

In [8]:
Sales_dataset.isna().sum()

pizza_id              0
order_id              0
pizza_name_id        16
quantity              0
order_date            0
order_time            0
unit_price            0
total_price           7
pizza_size            0
pizza_category       23
pizza_ingredients    13
pizza_name            7
dtype: int64

# Filling Total Price Based on Quantity:

In [9]:
Sales_dataset['total_price'].fillna(Sales_dataset['unit_price'] * Sales_dataset['quantity'], inplace=True)

# Filling Missing Pizza Categories Based on Pizza Name ID:

In [10]:
Sales_dataset['pizza_category'].value_counts()

pizza_category
Classic    14572
Supreme    11771
Veggie     11442
Chicken    10812
Name: count, dtype: int64

In [11]:
Sales_dataset[Sales_dataset['pizza_category'].isna()].head()

pizza_id  order_id   pizza_name_id  quantity order_date order_time  \
426       427       181      mexicana_l         1   3/1/2015   18:49:32   
427       428       182    brie_carre_s         1   3/1/2015   18:50:10   
428       429       182  peppr_salami_l         1   3/1/2015   18:50:10   
429       430       183    napolitana_m         1   3/1/2015   18:54:55   
430       431       183    spin_pesto_l         1   3/1/2015   18:54:55   

     unit_price  total_price pizza_size pizza_category  \
426       20.25        20.25          L            NaN   
427       23.65        23.65          S            NaN   
428       20.75        20.75          L            NaN   
429       16.00        16.00          M            NaN   
430       20.75        20.75          L            NaN   

                                     pizza_ingredients  \
426  Tomatoes, Red Peppers, Jalapeno Peppers, Red O...   
427  Brie Carre Cheese, Prosciutto, Caramelized Oni...   
428  Genoa Salami, Capocollo, Pepperoni, Tomatoes, ...   
429  Tomatoes, Anchovies, Green Olives, Red Onions,...   
430  Spinach, Artichokes, Tomatoes, Sun-dried Tomat...   

                  pizza_name  
426       The Mexicana Pizza  
427     The Brie Carre Pizza  
428  The Pepper Salami Pizza  
429     The Napolitana Pizza  
430  The Spinach Pesto Pizza

In [12]:
Sales_dataset['pizza_name_id'].value_counts()

pizza_name_id
big_meat_s        1811
thai_ckn_l        1363
five_cheese_l     1358
four_cheese_l     1272
classic_dlx_m     1159
                  ... 
mexicana_s         160
calabrese_s         99
ckn_alfredo_s       96
green_garden_l      94
the_greek_xxl       28
Name: count, Length: 91, dtype: int64

In [13]:
category_mapping = Sales_dataset[['pizza_name_id', 'pizza_category']].dropna().drop_duplicates()
category_mapping = category_mapping.set_index('pizza_name_id')['pizza_category'].to_dict()
Sales_dataset['pizza_category'] = Sales_dataset['pizza_category'].fillna(Sales_dataset['pizza_name_id'].map(category_mapping))

# Filling Missing Pizza Ingredients Based on Pizza Name:

In [14]:
Sales_dataset[Sales_dataset['pizza_ingredients'].isna()].head()

pizza_id  order_id   pizza_name_id  quantity order_date order_time  \
580       581       254   five_cheese_l         1   4/1/2015   22:22:13   
581       582       255   five_cheese_l         1   5/1/2015   11:23:35   
582       583       255  green_garden_m         1   5/1/2015   11:23:35   
583       584       255      hawaiian_l         1   5/1/2015   11:23:35   
584       585       255   spinach_fet_m         1   5/1/2015   11:23:35   

     unit_price  total_price pizza_size pizza_category pizza_ingredients  \
580        18.5         18.5          L         Veggie               NaN   
581        18.5         18.5          L         Veggie               NaN   
582        16.0         16.0          M         Veggie               NaN   
583        16.5         16.5          L        Classic               NaN   
584        16.0         16.0          M         Veggie               NaN   

                     pizza_name  
580       The Five Cheese Pizza  
581       The Five Cheese Pizza  
582      The Green Garden Pizza  
583          The Hawaiian Pizza  
584  The Spinach and Feta Pizza

In [15]:
ingredients_mapping = Sales_dataset[['pizza_name', 'pizza_ingredients']].dropna().drop_duplicates()
ingredients_mapping = ingredients_mapping.set_index('pizza_name')['pizza_ingredients'].to_dict()
Sales_dataset['pizza_ingredients'] = Sales_dataset['pizza_ingredients'].fillna(Sales_dataset['pizza_name'].map(ingredients_mapping))

# Filling Missing Pizza Name Based on Pizza Ingredients:

In [16]:
Sales_dataset[Sales_dataset['pizza_name'].isna()]

pizza_id  order_id pizza_name_id  quantity order_date order_time  \
688       689       303   pepperoni_s         1   5/1/2015   20:02:51   
689       690       304    hawaiian_s         1   5/1/2015   20:07:58   
690       691       304    mexicana_l         1   5/1/2015   20:07:58   
691       692       305   pepperoni_l         1   5/1/2015   20:35:40   
692       693       305    sicilian_s         1   5/1/2015   20:35:40   
693       694       305  spicy_ital_m         1   5/1/2015   20:35:40   
694       695       306    big_meat_s         1   5/1/2015   20:37:40   

     unit_price  total_price pizza_size pizza_category  \
688        9.75         9.75          S        Classic   
689       10.50        10.50          S        Classic   
690       20.25        20.25          L         Veggie   
691       15.25        15.25          L        Classic   
692       12.25        12.25          S        Supreme   
693       16.50        16.50          M        Supreme   
694       12.00        12.00          S        Classic   

                                     pizza_ingredients pizza_name  
688                       Mozzarella Cheese, Pepperoni        NaN  
689           Sliced Ham, Pineapple, Mozzarella Cheese        NaN  
690  Tomatoes, Red Peppers, Jalapeno Peppers, Red O...        NaN  
691                       Mozzarella Cheese, Pepperoni        NaN  
692  Coarse Sicilian Salami, Tomatoes, Green Olives...        NaN  
693  Capocollo, Tomatoes, Goat Cheese, Artichokes, ...        NaN  
694  Bacon, Pepperoni, Italian Sausage, Chorizo Sau...        NaN

In [17]:
ingredients_name_mapping = Sales_dataset[['pizza_ingredients', 'pizza_name']].dropna().drop_duplicates()
ingredients_name_mapping = ingredients_name_mapping.set_index('pizza_ingredients')['pizza_name'].to_dict()
Sales_dataset['pizza_name'] = Sales_dataset['pizza_name'].fillna(Sales_dataset['pizza_ingredients'].map(ingredients_name_mapping))

# Filling Missing Pizza Name ID Based on Pizza Name:

In [18]:
name_id_mapping = Sales_dataset[['pizza_name', 'pizza_name_id']].dropna().drop_duplicates()
name_id_mapping = name_id_mapping.set_index('pizza_name')['pizza_name_id'].to_dict()
Sales_dataset['pizza_name_id'] = Sales_dataset['pizza_name_id'].fillna(Sales_dataset['pizza_name'].map(name_id_mapping))

In [19]:
Sales_dataset.isna().sum()

pizza_id             0
order_id             0
pizza_name_id        0
quantity             0
order_date           0
order_time           0
unit_price           0
total_price          0
pizza_size           0
pizza_category       0
pizza_ingredients    0
pizza_name           0
dtype: int64

# Ingredients Data

In [20]:
Ingredients_dataset.head()

pizza_name_id                  pizza_name  pizza_ingredients  \
0     bbq_ckn_l  The Barbecue Chicken Pizza  Barbecued Chicken   
1     bbq_ckn_l  The Barbecue Chicken Pizza        Red Peppers   
2     bbq_ckn_l  The Barbecue Chicken Pizza      Green Peppers   
3     bbq_ckn_l  The Barbecue Chicken Pizza           Tomatoes   
4     bbq_ckn_l  The Barbecue Chicken Pizza         Red Onions   

   Items_Qty_In_Grams  
0                40.0  
1                15.0  
2                20.0  
3                30.0  
4                60.0

In [21]:
Ingredients_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 518 entries, 0 to 517
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   pizza_name_id       518 non-null    object 
 1   pizza_name          518 non-null    object 
 2   pizza_ingredients   518 non-null    object 
 3   Items_Qty_In_Grams  514 non-null    float64
dtypes: float64(1), object(3)
memory usage: 16.3+ KB


In [22]:
Ingredients_dataset.describe()

Items_Qty_In_Grams
count          514.000000
mean            30.564202
std             30.351623
min              5.000000
25%             10.000000
50%             20.000000
75%             40.000000
max            200.000000

# 4. Handling Missing Values in Ingredients Data:

In [23]:
Ingredients_dataset.isna().sum()

pizza_name_id         0
pizza_name            0
pizza_ingredients     0
Items_Qty_In_Grams    4
dtype: int64

In [24]:
Ingredients_dataset[Ingredients_dataset['Items_Qty_In_Grams'].isna()]

pizza_name_id            pizza_name   pizza_ingredients  \
24   brie_carre_s  The Brie Carre Pizza  Caramelized Onions   
149    hawaiian_l    The Hawaiian Pizza          Sliced Ham   
152    hawaiian_m    The Hawaiian Pizza          Sliced Ham   
155    hawaiian_s    The Hawaiian Pizza          Sliced Ham   

     Items_Qty_In_Grams  
24                  NaN  
149                 NaN  
152                 NaN  
155                 NaN

In [25]:
Ingredients_dataset[Ingredients_dataset['pizza_name']=='The Hawaiian Pizza']

pizza_name_id          pizza_name  pizza_ingredients  Items_Qty_In_Grams
149    hawaiian_l  The Hawaiian Pizza         Sliced Ham                 NaN
150    hawaiian_l  The Hawaiian Pizza          Pineapple                45.0
151    hawaiian_l  The Hawaiian Pizza  Mozzarella Cheese                30.0
152    hawaiian_m  The Hawaiian Pizza         Sliced Ham                 NaN
153    hawaiian_m  The Hawaiian Pizza          Pineapple                30.0
154    hawaiian_m  The Hawaiian Pizza  Mozzarella Cheese                20.0
155    hawaiian_s  The Hawaiian Pizza         Sliced Ham                 NaN
156    hawaiian_s  The Hawaiian Pizza          Pineapple                15.0
157    hawaiian_s  The Hawaiian Pizza  Mozzarella Cheese                10.0

In [26]:
mean_qty = Ingredients_dataset.groupby('pizza_name_id')['Items_Qty_In_Grams'].mean()

# Filling missing Items Quantity values using the Mean value:

In [27]:
Ingredients_dataset['Items_Qty_In_Grams'] = Ingredients_dataset['Items_Qty_In_Grams'].fillna(Ingredients_dataset['pizza_name_id'].map(mean_qty))

In [53]:
Ingredients_dataset.isna().sum()

pizza_name_id         0
pizza_name            0
pizza_ingredients     0
Items_Qty_In_Grams    0
dtype: int64

In [54]:
## save into CSV
Ingredients_dataset.to_csv("Ingredients_dataset.csv", index=False)

# 5. Merging Sales and Ingredients DataFrames:

In [28]:
merged_data = pd.merge(Sales_dataset, Ingredients_dataset, on='pizza_name_id', how='left')

In [29]:
merged_data.head()

pizza_id  order_id  pizza_name_id  quantity order_date order_time  \
0         1         1     hawaiian_m         1   1/1/2015   11:38:36   
1         1         1     hawaiian_m         1   1/1/2015   11:38:36   
2         1         1     hawaiian_m         1   1/1/2015   11:38:36   
3         2         2  classic_dlx_m         1   1/1/2015   11:57:40   
4         2         2  classic_dlx_m         1   1/1/2015   11:57:40   

   unit_price  total_price pizza_size pizza_category  \
0       13.25        13.25          M        Classic   
1       13.25        13.25          M        Classic   
2       13.25        13.25          M        Classic   
3       16.00        16.00          M        Classic   
4       16.00        16.00          M        Classic   

                                 pizza_ingredients_x  \
0           Sliced Ham, Pineapple, Mozzarella Cheese   
1           Sliced Ham, Pineapple, Mozzarella Cheese   
2           Sliced Ham, Pineapple, Mozzarella Cheese   
3  Pepperoni, Mushrooms, Red Onions, Red Peppers,...   
4  Pepperoni, Mushrooms, Red Onions, Red Peppers,...   

               pizza_name_x              pizza_name_y pizza_ingredients_y  \
0        The Hawaiian Pizza        The Hawaiian Pizza          Sliced Ham   
1        The Hawaiian Pizza        The Hawaiian Pizza           Pineapple   
2        The Hawaiian Pizza        The Hawaiian Pizza   Mozzarella Cheese   
3  The Classic Deluxe Pizza  The Classic Deluxe Pizza           Pepperoni   
4  The Classic Deluxe Pizza  The Classic Deluxe Pizza           Mushrooms   

   Items_Qty_In_Grams  
0                25.0  
1                30.0  
2                20.0  
3                40.0  
4                40.0

In [30]:
merged_data.shape

(267576, 15)

In [31]:
merged_data.isnull().sum()

pizza_id               0
order_id               0
pizza_name_id          0
quantity               0
order_date             0
order_time             0
unit_price             0
total_price            0
pizza_size             0
pizza_category         0
pizza_ingredients_x    0
pizza_name_x           0
pizza_name_y           0
pizza_ingredients_y    0
Items_Qty_In_Grams     0
dtype: int64

In [32]:
merged_data[['pizza_ingredients_x','pizza_ingredients_y']]

pizza_ingredients_x pizza_ingredients_y
0                Sliced Ham, Pineapple, Mozzarella Cheese          Sliced Ham
1                Sliced Ham, Pineapple, Mozzarella Cheese           Pineapple
2                Sliced Ham, Pineapple, Mozzarella Cheese   Mozzarella Cheese
3       Pepperoni, Mushrooms, Red Onions, Red Peppers,...           Pepperoni
4       Pepperoni, Mushrooms, Red Onions, Red Peppers,...           Mushrooms
...                                                   ...                 ...
267571  Barbecued Chicken, Red Peppers, Green Peppers,...         Red Peppers
267572  Barbecued Chicken, Red Peppers, Green Peppers,...       Green Peppers
267573  Barbecued Chicken, Red Peppers, Green Peppers,...            Tomatoes
267574  Barbecued Chicken, Red Peppers, Green Peppers,...          Red Onions
267575  Barbecued Chicken, Red Peppers, Green Peppers,...      Barbecue Sauce

[267576 rows x 2 columns]

In [33]:
merged_data[['pizza_name_x','pizza_name_y']]

pizza_name_x                pizza_name_y
0               The Hawaiian Pizza          The Hawaiian Pizza
1               The Hawaiian Pizza          The Hawaiian Pizza
2               The Hawaiian Pizza          The Hawaiian Pizza
3         The Classic Deluxe Pizza    The Classic Deluxe Pizza
4         The Classic Deluxe Pizza    The Classic Deluxe Pizza
...                            ...                         ...
267571  The Barbecue Chicken Pizza  The Barbecue Chicken Pizza
267572  The Barbecue Chicken Pizza  The Barbecue Chicken Pizza
267573  The Barbecue Chicken Pizza  The Barbecue Chicken Pizza
267574  The Barbecue Chicken Pizza  The Barbecue Chicken Pizza
267575  The Barbecue Chicken Pizza  The Barbecue Chicken Pizza

[267576 rows x 2 columns]

In [34]:
merged_data.drop(columns=['pizza_name_y', 'pizza_ingredients_y'], inplace=True)
merged_data.rename(columns={
    'pizza_name_x': 'pizza_name',
    'pizza_ingredients_x': 'pizza_ingredients'
}, inplace=True)

In [35]:
merged_data.columns

Index(['pizza_id', 'order_id', 'pizza_name_id', 'quantity', 'order_date',
       'order_time', 'unit_price', 'total_price', 'pizza_size',
       'pizza_category', 'pizza_ingredients', 'pizza_name',
       'Items_Qty_In_Grams'],
      dtype='object')

In [36]:
merged_data.isna().sum()

pizza_id              0
order_id              0
pizza_name_id         0
quantity              0
order_date            0
order_time            0
unit_price            0
total_price           0
pizza_size            0
pizza_category        0
pizza_ingredients     0
pizza_name            0
Items_Qty_In_Grams    0
dtype: int64

In [37]:
merged_data.duplicated().sum()

80204

In [38]:
# Droping duplicate rows from the dataset
merged_data = merged_data.drop_duplicates()

print(f"Number of rows after removing duplicates: {merged_data.shape[0]}")

Number of rows after removing duplicates: 187372


In [39]:
merged_data1 = merged_data.copy()

In [40]:
# Summary statistics
merged_data.describe()

pizza_id       order_id       quantity     unit_price  \
count  187372.000000  187372.000000  187372.000000  187372.000000   
mean    24312.676030   10702.312405       1.019250      16.914338   
std     14039.560378    6181.841091       0.141575       3.571451   
min         1.000000       1.000000       1.000000       9.750000   
25%     12166.750000    5344.000000       1.000000      12.750000   
50%     24266.000000   10661.000000       1.000000      16.750000   
75%     36494.250000   16112.250000       1.000000      20.500000   
max     48620.000000   21350.000000       4.000000      35.950000   

         total_price  Items_Qty_In_Grams  
count  187372.000000       187372.000000  
mean       17.244843           37.645900  
std         4.419979           35.729459  
min         9.750000            5.000000  
25%        12.750000           15.000000  
50%        16.750000           20.000000  
75%        20.750000           45.000000  
max        83.000000          200.000000

In [41]:
# Checking the data types and column names
merged_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 187372 entries, 0 to 267573
Data columns (total 13 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   pizza_id            187372 non-null  int64  
 1   order_id            187372 non-null  int64  
 2   pizza_name_id       187372 non-null  object 
 3   quantity            187372 non-null  int64  
 4   order_date          187372 non-null  object 
 5   order_time          187372 non-null  object 
 6   unit_price          187372 non-null  float64
 7   total_price         187372 non-null  float64
 8   pizza_size          187372 non-null  object 
 9   pizza_category      187372 non-null  object 
 10  pizza_ingredients   187372 non-null  object 
 11  pizza_name          187372 non-null  object 
 12  Items_Qty_In_Grams  187372 non-null  float64
dtypes: float64(3), int64(3), object(7)
memory usage: 20.0+ MB


# 6. Feature Exploration and Engineering

In [42]:
## Converting Order Dates to DateTime
def extract_date(date):
    formats = ['%d-%m-%Y', '%d/%m/%Y', '%m/%d/%Y']
    for fmt in formats:
        try:
            return pd.to_datetime(date, format=fmt)
        except ValueError:
            continue
    return pd.NaT

# Applying the function to the 'order_date' column
merged_data['order_date'] = merged_data['order_date'].apply(extract_date)

In [43]:
# Extracting the day of the week
merged_data['day_of_week'] = merged_data['order_date'].dt.dayofweek # day_name()
# Extracting the month
merged_data['month'] = merged_data['order_date'].dt.month # month_name()
# Createing a new column for the week number
merged_data['week'] = merged_data['order_date'].dt.isocalendar().week
# Extracting the year
merged_data['year'] = merged_data['order_date'].dt.year

# Adding Holiday Indicator to the Data

In [44]:
pip install holidays

In [45]:
import holidays

# using US holidays
us_holidays = holidays.US()

# Creating a 'holiday' column that indicates if the order_date was a holiday
merged_data['holiday'] = merged_data['order_date'].apply(lambda x: 1 if x in us_holidays else 0)

# Checking if the holiday flag works
merged_data[['order_date', 'holiday']].head()

order_date  holiday
0 2015-01-01        1
1 2015-01-01        1
2 2015-01-01        1
3 2015-01-01        1
6 2015-01-01        1

In [46]:
merged_data[['order_date', 'holiday']].tail()

order_date  holiday
267565 2015-12-31        0
267567 2015-12-31        0
267570 2015-12-31        0
267571 2015-12-31        0
267573 2015-12-31        0

In [47]:
# Adding Weekend Promotion Flag(assuming promotions happen on weekends)
# Creating a 'promotional_period' flag for weekends 
merged_data['promotion'] = merged_data['order_date'].dt.weekday.apply(lambda x: 1 if x >= 5 else 0)

# Checking if the promotion flag works
merged_data[['order_date', 'promotion']].head()

order_date  promotion
0 2015-01-01          0
1 2015-01-01          0
2 2015-01-01          0
3 2015-01-01          0
6 2015-01-01          0

In [48]:
merged_data[merged_data['promotion'] == 1][['order_date', 'promotion']].head()

order_date  promotion
1789 2015-01-03          1
1790 2015-01-03          1
1791 2015-01-03          1
1793 2015-01-03          1
1794 2015-01-03          1

In [49]:
merged_data.isnull().sum()

pizza_id              0
order_id              0
pizza_name_id         0
quantity              0
order_date            0
order_time            0
unit_price            0
total_price           0
pizza_size            0
pizza_category        0
pizza_ingredients     0
pizza_name            0
Items_Qty_In_Grams    0
day_of_week           0
month                 0
week                  0
year                  0
holiday               0
promotion             0
dtype: int64

In [50]:
merged_data.head()

pizza_id  order_id  pizza_name_id  quantity order_date order_time  \
0         1         1     hawaiian_m         1 2015-01-01   11:38:36   
1         1         1     hawaiian_m         1 2015-01-01   11:38:36   
2         1         1     hawaiian_m         1 2015-01-01   11:38:36   
3         2         2  classic_dlx_m         1 2015-01-01   11:57:40   
6         2         2  classic_dlx_m         1 2015-01-01   11:57:40   

   unit_price  total_price pizza_size pizza_category  \
0       13.25        13.25          M        Classic   
1       13.25        13.25          M        Classic   
2       13.25        13.25          M        Classic   
3       16.00        16.00          M        Classic   
6       16.00        16.00          M        Classic   

                                   pizza_ingredients  \
0           Sliced Ham, Pineapple, Mozzarella Cheese   
1           Sliced Ham, Pineapple, Mozzarella Cheese   
2           Sliced Ham, Pineapple, Mozzarella Cheese   
3  Pepperoni, Mushrooms, Red Onions, Red Peppers,...   
6  Pepperoni, Mushrooms, Red Onions, Red Peppers,...   

                 pizza_name  Items_Qty_In_Grams  day_of_week  month  week  \
0        The Hawaiian Pizza                25.0            3      1     1   
1        The Hawaiian Pizza                30.0            3      1     1   
2        The Hawaiian Pizza                20.0            3      1     1   
3  The Classic Deluxe Pizza                40.0            3      1     1   
6  The Classic Deluxe Pizza                10.0            3      1     1   

   year  holiday  promotion  
0  2015        1          0  
1  2015        1          0  
2  2015        1          0  
3  2015        1          0  
6  2015        1          0

In [51]:
# Aggregateing sales data by date
daily_sales = merged_data.groupby('order_date')['quantity'].sum().reset_index()
# daily_sales.set_index('order_date', inplace=True)
daily_sales

order_date  quantity
0   2015-01-01       620
1   2015-01-02       631
2   2015-01-03       618
3   2015-01-04       414
4   2015-01-05       482
..         ...       ...
353 2015-12-27       347
354 2015-12-28       381
355 2015-12-29       302
356 2015-12-30       327
357 2015-12-31       692

[358 rows x 2 columns]

In [52]:
## save into CSV
#merged_data.to_csv("MERGED_DOMINOS_DATA.csv", index=False)